unittests/merkle_tree_tests.cpp

```diff
@@ -0,0 +1,263 @@
+#include <eosio/chain/incremental_merkle.hpp>
+#include <boost/test/unit_test.hpp>
+#include <fc/crypto/sha256.hpp>
+
+using namespace eosio::chain;
+
+BOOST_AUTO_TEST_SUITE(merkle_tree_tests)
+
+BOOST_AUTO_TEST_CASE(basic_append_and_root_check_canonical) {
+   incremental_canonical_merkle tree;
+   BOOST_CHECK_EQUAL(tree.get_root(), fc::sha256());
+
+   auto node1 = fc::sha256::hash("Node1");
+   tree.append(node1);
+   BOOST_CHECK_EQUAL(tree.get_root(), node1);
+}
+
+BOOST_AUTO_TEST_CASE(multiple_appends_canonical) {
+   incremental_canonical_merkle tree;
+   auto node1 = fc::sha256::hash("Node1");
+   auto node2 = fc::sha256::hash("Node2");
+   auto node3 = fc::sha256::hash("Node3");
+   auto node4 = fc::sha256::hash("Node4");
+   auto node5 = fc::sha256::hash("Node5");
+   auto node6 = fc::sha256::hash("Node6");
+   auto node7 = fc::sha256::hash("Node7");
+   auto node8 = fc::sha256::hash("Node8");
+   auto node9 = fc::sha256::hash("Node9");
+
+   tree.append(node1);
+   BOOST_CHECK_EQUAL(tree.get_root(), node1);
+
+   tree.append(node2);
+   BOOST_CHECK_EQUAL(tree.get_root().str(), fc::sha256::hash(make_canonical_pair(node1, node2)).str());
```

Comment:
why `.str()` here and below, and not on the previous one?

Reply:
The str() gives a readable error message.

Reply:
So the one at line 31 should also have a `str()`, right?